cmd/containerd-stress/cri_worker.go

```diff
@@ -119,22 +119,22 @@ func (w *criWorker) runSandbox(tctx, ctx context.Context, id string) (err error)
 
 	ticker := time.NewTicker(250 * time.Millisecond)
 	quit := make(chan struct{})
+
 	go func() {
+		defer func() {
+			ticker.Stop()
+			close(quit)
+		}()
+
 		for {
 			select {
 			case <-tctx.Done():
-				close(quit)
 				return
 			case <-ticker.C:
-				// do stuff
 				status, err := w.client.PodSandboxStatus(sb)
-				if err != nil && status.GetState() == runtime.PodSandboxState_SANDBOX_READY {
-					close(quit)
+				if err == nil && status.GetState() == runtime.PodSandboxState_SANDBOX_READY {
 					return
 				}
-			case <-quit:
-				ticker.Stop()
-				return
 			}
 		}
 	}()
```

Comment:
By logic of the modification, the quit channel don't make any sense.
And Tricker will not be referenced again when the go func exits, and will then be garbage collected.
```go
go func() {
  for {
      select {
      case <- tctx.Done():
          return
      case <-ticker.C:
          status, err := w.client.PodSandboxStatus(sb)
          if err == nil && status.GetState() == runtime.PodSandboxState_SANDBOX_READY {
            return
          }
      }
  }
}()
```

**But when we carefully analyze the contextual logic, we find that this function seems to be missing the `<- quit` at the end**

Reply:
ye, got it, Maybe we should use WaitGroup

```go
ticker := time.NewTicker(250 * time.Millisecond)

	wg := sync.WaitGroup{}
	wg.Add(1)
	go func() {
		defer ticker.Stop()
		defer wg.Done()
		for {
			select {
			case <-tctx.Done():
				return
			case <-ticker.C:
				status, err := w.client.PodSandboxStatus(sb)
				if err == nil && status.GetState() == runtime.PodSandboxState_SANDBOX_READY {
					return
				}
			}
		}
	}()

	wg.Wait()
```